src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributeList.cs

```diff
@@ -2,73 +2,52 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
 
 namespace Microsoft.AspNetCore.Razor.TagHelpers
 {
     /// <summary>
-    /// A read-only collection of <typeparamref name="TAttribute"/>s.
+    /// A read-only collection of <typeparamref name="TagHelperAttribute"/>s.
     /// </summary>
-    /// <typeparam name="TAttribute">
-    /// The type of <see cref="IReadOnlyTagHelperAttribute"/>s in the collection.
+    /// <typeparam name="TagHelperAttribute">
+    /// The type of <see cref="TagHelperAttribute"/>s in the collection.
     /// </typeparam>
-    public class ReadOnlyTagHelperAttributeList<TAttribute> : IReadOnlyList<TAttribute>
-        where TAttribute : IReadOnlyTagHelperAttribute
+    public class ReadOnlyTagHelperAttributeList : ReadOnlyCollection<TagHelperAttribute>
     {
-        private static readonly IReadOnlyList<TAttribute> EmptyList =
-#if NET451
-            new TAttribute[0];
-#else
-            Array.Empty<TAttribute>();
-#endif
+        private static readonly IReadOnlyList<TagHelperAttribute> EmptyList = new TagHelperAttribute[0];
 
         /// <summary>
-        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> with an empty
+        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList"/> with an empty
         /// collection.
         /// </summary>
         protected ReadOnlyTagHelperAttributeList()
+            : base(new List<TagHelperAttribute>())
         {
-            Attributes = new List<TAttribute>();
         }
 
         /// <summary>
-        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> with the specified
+        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList"/> with the specified
         /// <paramref name="attributes"/>.
         /// </summary>
         /// <param name="attributes">The collection to wrap.</param>
-        public ReadOnlyTagHelperAttributeList(IEnumerable<TAttribute> attributes)
+        public ReadOnlyTagHelperAttributeList(IEnumerable<TagHelperAttribute> attributes)
+            : base(new List<TagHelperAttribute>(attributes))
```

Comment:
Is this constructor used in `ReadOnlyTagHelperAttributeList` copy scenarios? The defensive copy of the given collection (which may already be an instance of this type) looks wasteful, especially if callers have real collections and not just `IEnumerable<T>`s.